cpp/src/arrow/filesystem/filesystem.cc

```diff
@@ -738,6 +761,109 @@ Result<std::shared_ptr<FileSystem>> FileSystemFromUriReal(const Uri& uri,
 
 }  // namespace
 
+Status RegisterFileSystemFactory(std::vector<std::string> schemes,
+                                 FileSystem::Factory factory) {
+  auto& [mutex, scheme_to_factory] = FileSystemFactoryRegistry();
+  std::unique_lock lock{mutex};
+
+  for (auto&& scheme : schemes) {
+    if (auto& ref = scheme_to_factory[scheme]) {
+      return Status::KeyError(
+          "Tried to add a factory for ", scheme,
+          ":// uris, but a factory was already registered for that scheme");
+    } else {
+      ref = factory;
+    }
+  }
+  return Status::OK();
+}
+
+// XXX alternative approach:
+// If we distrust global constructors more than we distrust
+// symbol visibility, we could export specially named functions
+// and retrieve them via dlsym.
+Status RegisterFileSystemFactoryModule(std::string module_name) {
+  static auto* handle = dlopen(nullptr, RTLD_NOW | RTLD_LOCAL);
+  if (!handle) {
+    return Status::Invalid("dlopen failed: ", dlerror());
+  }
+
+  dlerror();
+
+  module_name = "ArrowFileSystemModule_" + module_name;
```

Comment:
I don't think we want to impose a naming convention here, why would we?

Reply:
If we use extern "C" module functions, the prefixed naming convention allows us to support linking to (and not only dynamically loading) libraries containing modules. One pattern for dlopen-only modules would be:
- dlopen the library `libarrow_fs_s3.so`
- dlsym for "hardcoded_module_function"
- invoke that function pointer
(this is the basic format for [node addons )) However if we then link to `libarrow_fs_s3.so` and `libarrow_fs_hdfs.so`, then we'll get a link error because "hardcoded_module_function" has multiple definitions. If that's not something we'd ever do (and we decide to use extern "C" module functions) then it'd certainly be less ugly to avoid the naming convention.

Reply:
Before discussing technical solutions, please let's decide what our needs are here.

Reply:
What API should we provide for C++ users and C++ bindings users (such as PyArrow users)?

I think that it's convenient that `FileSystemFromUri("s3://...")` loads `S3FileSystem` automatically. C++ bindings users don't need to change anything to support this API because they already have `FileSystemFromUri()` bindings. But it's difficult to finalize automatically loaded file system implementations... For example, users can't call `arrow::fs::EnsureS3Finalized()` with this API...

Reply:
> I think that it's convenient that `FileSystemFromUri("s3://...")` loads `S3FileSystem` automatically.

I agree that it _would be_ convenient, but how do you plan to ensure that works on all distributions of Arrow? Unlike other languages, in C++ we don't control how Arrow will be installed or loaded by third-party code. Some people might even want to change the file names (for example `libmyvendored_arrow` instead of `libarrow`).

> For example, users can't call `arrow::fs::EnsureS3Finalized()` with this API

Is calling `EnsureS3Finalized` still required?

Reply:
> What API should we provide for C++ users and C++ bindings users (such as PyArrow users)?

From PyArrow's point of view, I think it is completely fine that we, as pyarrow, first have to explicitly register the S3FileSystem before `FileSystemFromUri("s3://...")` works. 
PyArrow could check, on first usage of the filesystems, which known filesystem packages are installed and register them all. And for truly custom (third party) filesystems, users will just have to import such package first before URI strings are recognized automatically. I think it is fine to expect that.

Reply:
> > I think that it's convenient that `FileSystemFromUri("s3://...")` loads `S3FileSystem` automatically.
> 
> I agree that it _would be_ convenient, but how do you plan to ensure that works on all distributions of Arrow? Unlike other languages, in C++ we don't control how Arrow will be installed or loaded by third-party code. Some people might even want to change the file names (for example `libmyvendored_arrow` instead of `libarrow`).

How about providing a configuration to change base name like jemalloc does? (See `--with-install-suffix` in https://raw.githubusercontent.com/jemalloc/jemalloc/dev/INSTALL.md )
If we provide a configuration for it and recommend it, we can know installed library name.

But this will not work with static library only build. In the case, users need to embed file system implementations they need to their binary.
 
> > For example, users can't call `arrow::fs::EnsureS3Finalized()` with this API
> 
> Is calling `EnsureS3Finalized` still required?

I think so because we still have this warning: https://github.com/apache/arrow/blob/4b1f06327f05341b6e51293b3186d80cd5fdbf87/cpp/src/arrow/filesystem/s3fs.cc#L2913-L2917
If we don't need to call `EnsureS3Finalized` explicitly, can we remove the warning?

Reply:
> How about providing a configuration to change base name like jemalloc does?

We can, but before doing that, we should provide the basic blocks so that users implement it themselves easily.

> I think so because we still have this warning:

I was misremembering. Thanks.

Reply:
> I think that it's convenient that FileSystemFromUri("s3://...") loads S3FileSystem automatically.

This can be accomplished for built-in filesystems as long as `libarrow.so` can derive the location and name of `libarrow_fs_s3.so`. However I think it's inevitable that some installations will render these libraries difficult to find, in which case the paths will need to be explicitly provided to `libarrow.so` by the consumer.

> But it's difficult to finalize automatically loaded file system implementations... For example, users can't call arrow::fs::EnsureS3Finalized() with this API...

Initialization of filesystems can be handled in the factory on the first call. Part of this initialization could be appending the implementation's finalizer to a list, all of which are invoked by `arrow::fs::Finalize()`. I think this is actually cleaner than requiring users to individually finalize each filesystem they use.

> But this will not work with static library only build.

One advantage of using global constructors (registrars) is that as long as the binary which contains the definition is loaded, the factories will be registered with no further intervention by the consumer. If the S3 filesystem is dynamically loaded then the factory will be registered before `dlopen()` returns. Otherwise (for example if `libarrow_fs_s3.a` is statically linked to the consumer) then the factory will be registered before `main()` is entered.

Reply:
> > I think that it's convenient that FileSystemFromUri("s3://...") loads S3FileSystem automatically.
> 
> This can be accomplished for built-in filesystems as long as `libarrow.so` can derive the location and name of `libarrow_fs_s3.so`. However I think it's inevitable that some installations will render these libraries difficult to find, in which case the paths will need to be explicitly provided to `libarrow.so` by the consumer.

I agree with you. I think that we don't need to cover all use cases without explicit consumer configuration. It's enough that we can cover most use cases. If we provide a naming convention or something, most use cases may  consider it.
 
> > But it's difficult to finalize automatically loaded file system implementations... For example, users can't call arrow::fs::EnsureS3Finalized() with this API...
> 
> Initialization of filesystems can be handled in the factory on the first call. Part of this initialization could be appending the implementation's finalizer to a list, all of which are invoked by `arrow::fs::Finalize()`. I think this is actually cleaner than requiring users to individually finalize each filesystem they use.

Wow. I didn't notice the approach. Thanks.

> > But this will not work with static library only build.
> 
> One advantage of using global constructors (registrars) is that as long as the binary which contains the definition is loaded, the factories will be registered with no further intervention by the consumer. If the S3 filesystem is dynamically loaded then the factory will be registered before `dlopen()` returns. Otherwise (for example if `libarrow_fs_s3.a` is statically linked to the consumer) then the factory will be registered before `main()` is entered.

I agree with you.